src/core/frames/frame_controller.ts

```diff
@@ -145,39 +134,30 @@ export class FrameController
 
   private async loadSourceURL() {
     if (this.enabled && this.isActive && !this.complete && this.sourceURL) {
-      this.element.loaded = this.visit(expandURL(this.sourceURL))
-      this.appearanceObserver.stop()
-      await this.element.loaded
-      this.hasBeenLoaded = true
+      await this.visit({ url: this.sourceURL })
     }
   }
 
-  async loadResponse(fetchResponse: FetchResponse) {
+  async loadResponse(fetchResponse: FetchResponse, frameVisit: FrameVisit) {
     if (fetchResponse.redirected || (fetchResponse.succeeded && fetchResponse.isHTML)) {
       this.sourceURL = fetchResponse.response.url
     }
 
-    try {
-      const html = await fetchResponse.responseHTML
-      if (html) {
-        const document = parseHTMLDocument(html)
-        const pageSnapshot = PageSnapshot.fromDocument(document)
-
-        if (pageSnapshot.isVisitable) {
-          await this.loadFrameResponse(fetchResponse, document)
-        } else {
-          await this.handleUnvisitableFrameResponse(fetchResponse)
-        }
+    const html = await fetchResponse.responseHTML
+    if (html) {
+      const pageSnapshot = PageSnapshot.fromHTMLString(html)
+
+      if (pageSnapshot.isVisitable) {
+        await this.loadFrameResponse(fetchResponse, pageSnapshot, frameVisit)
```

Comment:
@kevinmcconnell @afcapel I've rebased this commit to incorporate the changes made in https://github.com/hotwired/turbo/pull/867, namely the expectation that [there will be a `PageSnapshot` instance at-the-ready](https://github.com/hotwired/turbo/commit/1e78f3b1ef4e5263c4d6fd2003ae8298e27b3b3c#diff-8f44fafd854f6f716d51a2b05b5689266a9b66e0b1ad9d68570bd4e8cefd0a47).